pxtwinrt/serial.ts

```diff
@@ -1,51 +1,84 @@
+    let deviceNameFilter: RegExp;
+    let activePorts: pxt.Map<DevicePort> = {}
 
     interface DevicePort {
-        info: any;
-        device?: any;
+        info: DeviceInfo;
+        device?: SerialDevice;
     }
 
     export function initSerial() {
         if (!pxt.appTarget.serial
             || !pxt.appTarget.serial.log
             || !pxt.appTarget.serial.nameFilter) return;
 
-        const filter = new RegExp(pxt.appTarget.serial.nameFilter);
-        const serialDeviceSelector = (Windows.Devices as any).SerialCommunication.SerialDevice.getDeviceSelector();
-        // Create a device watcher to look for instances of the Serial device
-        // The createWatcher() takes a string only when you provide it two arguments, so be sure to include an array as a second
-        // parameter (JavaScript can only recognize overloaded functions with different numbers of parameters).
-        watcher = Windows.Devices.Enumeration.DeviceInformation.createWatcher(serialDeviceSelector, [] as any);
-        watcher.addEventListener("added", (dis: any) => {
-            toArray(dis.detail).forEach((di: any) => {
-                if (!filter.test(di.name)) return;
-                pxt.debug(`serial port added ${di.name} - ${di.id}`);
-                ports[di.id] = {
-                    info: di
-                };
-                (Windows.Devices as any).SerialCommunication.SerialDevice.fromIdAsync(di.id)
-                    .done((dev: any) => {
-                        ports[di.id].device = dev;
-                        startDevice(di.id);
-                    })
-            });
-        });
-        watcher.addEventListener("removed", (dis: any) => {
-            toArray(dis.detail).forEach((di: any) => delete ports[di.id]);
-        })
-        watcher.addEventListener("updated", (dis: any) => {
-            toArray(dis.detail).forEach((di: any) => ports[di.id] ? ports[di.id].info.update(di.info) : null);
-        })
+        if (!watcher) {
+            deviceNameFilter = new RegExp(pxt.appTarget.serial.nameFilter);
+            const serialDeviceSelector = Windows.Devices.SerialCommunication.SerialDevice.getDeviceSelector();
+            // Create a device watcher to look for instances of the Serial device
+            // As per MSDN doc, to use the correct overload, we pass null as 2nd argument
+            watcher = Windows.Devices.Enumeration.DeviceInformation.createWatcher(serialDeviceSelector, null);
+        }
+
+        watcher.addEventListener("added", deviceAdded);
+        watcher.addEventListener("removed", deviceRemoved);
+        watcher.addEventListener("updated", deviceUpdated);
         watcher.start();
     }
 
+    export function suspendSerial() {
+        if (watcher) {
+            watcher.stop();
+            watcher.removeEventListener("added", deviceAdded);
+            watcher.removeEventListener("removed", deviceRemoved);
+            watcher.removeEventListener("updated", deviceUpdated);
```

Comment:
set watcher to undefined?

Reply:
According to the docs, `stop()` resets the state, so I was only constructing it the first time and restarting it on `resume`. But I guess it's safer to just delete it and use a new one, so I'll do that